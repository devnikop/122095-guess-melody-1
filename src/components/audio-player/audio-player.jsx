import React from 'react';
import propTypes from 'prop-types';

export class AudioPlayer extends React.PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      progress: 0,
      idLoading: true,
      isPlaying: props.isPlaying,
    };

    this._onPlayButtonClick = this._onPlayButtonClick.bind(this);
  }

  render() {
    const {isLoading, isPlaying} = this.state;

    return <div className="game__track">
      <button
        className={`track__button track__button--${isPlaying ? `pause` : `play`}`}
        type="button"
        disabled={isLoading}
        onClick={this._onPlayButtonClick}
      />
      <div className="track__status">
        <audio/>
      </div>
    </div>;
  }

  componentDidMount() {
    this._audio = new Audio(this.props.src);

    this._audio.oncanplaythrough = () =>
      this.setState({
        isLoading: false,
      });

    this._audio.onplay = () => {
      this.setState({
        isPlaying: true,
      });
    };

    this._audio.onpause = () =>
      this.setState({
        isPlaying: false,
      });

    this._audio.ontimeupdate = () =>
      this.setState({
        progress: this._audio.currentTime
      });
  }

  componentDidUpdate() {
    if (this.props.isPlaying) {
      this._audio.play();
    } else {
      this._audio.pause();
    }
  }

  componentWillUnmount() {
    this._audio.oncanplaythrough = null;
    this._audio.onplay = null;
    this._audio.onpause = null;
    this._audio.ontimeupdate = null;
    this._audio.src = ``;
    this._audio = null;
  }

  _onPlayButtonClick() {
    this.props.onPlayButtonClick();
    this.setState({isPlaying: !this.state.isPlaying});
  }
}

AudioPlayer.propTypes = {
  isPlaying: propTypes.bool.isRequired,
  onPlayButtonClick: propTypes.func.isRequired,
  src: propTypes.string.isRequired,
};
