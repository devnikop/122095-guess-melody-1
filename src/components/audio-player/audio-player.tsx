import * as React from 'react';

interface Props {
  isLoading: boolean,
  isPlaying: boolean,
  onPlayButtonClick: () => void,
  renderAudio: () => React.ReactElement,
}

export class AudioPlayer extends React.PureComponent<Props> {
  render() {
    const {
      isLoading,
      isPlaying,
      renderAudio,
      onPlayButtonClick,
    } = this.props;

    return <div className="game__track">
      <button
        className={`
          track__button
          track__button--${isPlaying ? `pause` : `play`}
        `}
        type="button"
        disabled={isLoading}
        onClick={onPlayButtonClick}
      />
      <div className="track__status">
        {renderAudio()}
      </div>
    </div>;
  }
}
