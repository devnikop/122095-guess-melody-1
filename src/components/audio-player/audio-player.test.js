import React from 'react';
import renderer from 'react-test-renderer';

import {AudioPlayer} from './audio-player.jsx';

const mock = {
  src: `https://upload.wikimedia.org/wikipedia/commons/1/1f/Uganda_flag_and_national_anthem_-_Oh_Uganda_Land_o.ogg`,
};

it(`AudioPlayer correctly renders`, () => {
  const {src} = mock;

  const tree = renderer
    .create(<AudioPlayer
      isPlaying={true}
      isLoading={true}
      onPlayButtonClick={jest.fn()}
      renderAudio={jest.fn()}
      src={src}
    />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});
