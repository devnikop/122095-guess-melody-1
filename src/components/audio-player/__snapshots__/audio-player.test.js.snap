// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AudioPlayer correctly renders 1`] = `
<div
  className="game__track"
>
  <button
    className="track__button track__button--pause"
    onClick={[Function]}
    type="button"
  />
  <div
    className="track__status"
  >
    <audio />
  </div>
</div>
`;
